presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java

```diff
@@ -701,13 +701,26 @@ public PlanNode visitValues(ValuesNode node, RewriteContext<Set<Symbol>> context
         @Override
         public PlanNode visitApply(ApplyNode node, RewriteContext<Set<Symbol>> context)
         {
-            ImmutableSet<Symbol> inputContext = ImmutableSet.<Symbol>builder()
+            PlanNode subquery = context.rewrite(node.getSubquery(), context.get());
+
+            // prune not used correlation symbols
+            Set<Symbol> subquerySymbols = DependencyExtractor.extractUnique(subquery);
+            List<Symbol> newCorrelation = node.getCorrelation().stream()
+                    .filter(subquerySymbols::contains)
+                    .collect(toImmutableList());
+
+            Set<Symbol> inputContext = ImmutableSet.<Symbol>builder()
                     .addAll(context.get())
-                    .addAll(node.getCorrelation())
+                    .addAll(newCorrelation)
                     .build();
             PlanNode input = context.rewrite(node.getInput(), inputContext);
-            PlanNode subquery = context.rewrite(node.getSubquery(), context.get());
-            return new ApplyNode(node.getId(), input, subquery, node.getCorrelation());
+
+            if (!inputContext.stream().anyMatch(subquery.getOutputSymbols()::contains)) {
+                // prune ApplyNode as subquery symbols are not used
```

Comment:
No.

This handles queries like: 
```
SELECT COUNT(*) FROM (
  SELECT orderkey IN (SELECT orderkey 
    FROM lineitem 
    WHERE orderkey % 2 = 0)
  FROM orders LIMIT 10)
```

Please notice that actual result of IN predicate is not used and so ApplyNode should be removed regardless how many rows has input relation.